src/Sema.zig

```diff
@@ -16445,17 +16455,71 @@ fn analyzeArithmetic(
     const rhs_zig_ty_tag = try rhs_ty.zigTypeTagOrPoison(mod);
     try sema.checkVectorizableBinaryOperands(block, src, lhs_ty, rhs_ty, lhs_src, rhs_src);
 
-    if (lhs_zig_ty_tag == .Pointer) switch (lhs_ty.ptrSize(mod)) {
-        .One, .Slice => {},
-        .Many, .C => {
-            const air_tag: Air.Inst.Tag = switch (zir_tag) {
-                .add => .ptr_add,
-                .sub => .ptr_sub,
-                else => return sema.fail(block, src, "invalid pointer arithmetic operator", .{}),
-            };
-            return sema.analyzePtrArithmetic(block, src, lhs, rhs, air_tag, lhs_src, rhs_src);
-        },
-    };
+    if (lhs_zig_ty_tag == .Pointer) {
+        if (rhs_zig_ty_tag == .Pointer) {
+            if (lhs_ty.ptrSize(mod) != .Slice and rhs_ty.ptrSize(mod) != .Slice) {
+                if (zir_tag != .sub) {
+                    return sema.failWithInvalidPtrArithmetic(block, src, "pointer-pointer", "subtraction");
+                }
+
+                const runtime_src = runtime_src: {
+                    if (try sema.resolveValue(lhs)) |lhs_value| {
+                        if (try sema.resolveValue(rhs)) |rhs_value| {
+                            const lhs_ptr = switch (mod.intern_pool.indexToKey(lhs_value.toIntern())) {
+                                .undef => return sema.failWithUseOfUndef(block, lhs_src),
+                                .ptr => |ptr| ptr,
+                                else => unreachable,
+                            };
+                            const rhs_ptr = switch (mod.intern_pool.indexToKey(rhs_value.toIntern())) {
+                                .undef => return sema.failWithUseOfUndef(block, rhs_src),
+                                .ptr => |ptr| ptr,
+                                else => unreachable,
+                            };
+                            // Make sure the pointers point to the same object.
+                            if (!lhs_ptr.base_addr.eql(rhs_ptr.base_addr)) {
+                                if (lhs_ptr.base_addr == .field) {
+                                    // In the case of a struct, make sure:
+                                    // 1. Both field pointers point to fields of the same struct by comparing only the bases of the field.
+                                    // The `eql` check above already makes sure both values are the same tag.
+                                    if (lhs_ptr.base_addr.field.base != rhs_ptr.base_addr.field.base) {
+                                        break :runtime_src src;
+                                    }
+                                    // 2. The struct of the fields that the two pointers point to is well-defined.
+                                    if (lhs_ptr.byte_offset == 0 and rhs_ptr.byte_offset == 0) {
```

Comment:
I'm not 100% sure about this but looking at the test cases that pass with this it seems to work to some extent.

Reply:
Based on `getOffsetPtr` existing I'd say it isn't. I think it'd be better to leave as it was with field pointers not working at comptime and make a follow-up issue for it.

The change to `hash_map.zig` could also be dropped to avoid having to update `zig1.wasm`.

Reply:
Yeah, unfortunately this is a bit tricky because it goes into how we're dealing with pointer provenance for auto-layout aggregates, which AFAIK is not yet decided.

To be clear, here's how the comptime pointer representation works:
* A pointer is represented by a chain of pointers terminating in one with `base_addr` of `int`, `decl`, `anon_decl`, `comptime_alloc`, or `comptime_field`.
* Values in this chain other than the terminal value represent accesses of auto-layout aggregates. For instance, if you get a pointer to a field of a normal `struct`, you get a pointer with `.base_addr = .{ .field = ... }`,
* Accesses to fields/elements of things with a well-defined layout instead just update `byte_offset`. This also applies to all arrays, since even if the element type has ill-defined layout, the ABI size of the type is always observable and has defined semantics. (The exception is arrays where the element type is comptime-only, since there is no meaningful byte offset for these.)

If we're going to define pointers to e.g. fields of auto-layout structs to have different provenance that pointers to other fields, then the correct thing is to error if any part of the pointer other than the final byte offset differs. So, if you have pointer chains like `field[0; offset=100], decl[offset=0]` and `field[1; offset=200], decl[offset=0]`, that errors, because the field indices `0` and `1` are differentl while `field[0; offset=100], decl[offset=0]` and `field[0; offset=228], decl[offset=0]` works and gives byte offset 128 (translated into elements appropriately). The only case where differing structures should be allowed in this case is for elements of comptime-only arrays.

OTOH, if we're going to be more lax about provenance in auto-layout aggregates, then... we might end up changing the pointer representation again to be honest, to use byte offsets in more cases.

Regardless, it's fine for now to have this not work at comptime and open a follow-up issue.

Reply:
Alright well with these changes reverted the same tests still pass so not even sure what to open an issue for except for maybe `struct` at comptime.

Reply:
Thanks for the detailed response! It looked like something like that would be the case but I wasn't familiar enough with the code to say for sure.